mount/mount.c

```diff
@@ -54,30 +45,223 @@ static int dns_lookup(const char *hostname, struct sockaddr_in *addr)
+}
+
+static inline int fsmount(int fsfd, unsigned int flags, unsigned int ms_flags)
+{
+	return syscall(__NR_fsmount, fsfd, flags, ms_flags);
+}
+
+static inline int fsconfig(int fsfd, unsigned int cmd,
+			   const char *key, const void *val, int aux)
+{
+	return syscall(__NR_fsconfig, fsfd, cmd, key, val, aux);
+}
+
+static inline int move_mount(int from_dfd, const char *from_pathname,
+			     int to_dfd, const char *to_pathname,
+			     unsigned int flags)
+{
+	return syscall(__NR_move_mount,
+		       from_dfd, from_pathname,
+		       to_dfd, to_pathname, flags);
+}
+
+#define E_fsconfig(fd, cmd, key, val, aux)                              \
+        do {                                                            \
+                if (fsconfig(fd, cmd, key, val, aux) == -1)             \
+                        mount_error(fd, key ?: "create");               \
+        } while (0)
+
+
+void process_option(char* option, int fsfd) {
+	char *key, *value, *saveptr;
+	key = strtok_r (option, "=", &saveptr);
+	option = NULL;
+	value = strtok_r (option, "=", &saveptr);
+	fprintf(stderr, "titus-mount: Setting filesystem mount option %s=%s\n", key, value);
+	E_fsconfig(fsfd, FSCONFIG_SET_STRING, key, value, 0);
+}
+
+void do_fsconfigs(int fsfd, char *options) {
+	char *str1, *token, *saveptr;
+	/* Mount options come in in the classic comma-separated key=value pairs
+	   we need to split them up and pass them in for fsconfig to handle one at a time */
+	for (str1 = options; ;str1 = NULL) {
+		token = strtok_r(str1, ",", &saveptr);
+		if (token == NULL)
+			break;
+		process_option(token, fsfd);
+	}
+	/* This last FSCONFIG_CMD_CREATE fsconfig call actually creates the superblock */
+	E_fsconfig(fsfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
+}
+
+
+static int setup_fsfd_in_namespaces(int sk, int pidfd) {
+	int fsfd, ret;
+	ret = setns(pidfd, CLONE_NEWUSER);
+	if (ret == -1) {
+		perror("setns user");
+		return 1;
+	}
+	ret = setns(pidfd, CLONE_NEWNET);
+	if (ret == -1) {
+		perror("setns net");
+		return 1;
+	}
+	ret = setns(pidfd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mnt");
+		return 1;
+	}
+	fsfd = fsopen("nfs4", 0);
+	if (fsfd == -1) {
+		perror("fsopen");
+		return 1;
+	}
+	assert(send_fd(sk, fsfd) == 0);
+	return 0;
+}
+
+int fork_and_get_fsfd(long int pidfd) {
+	int  sk_pair[2], ret, fsfd, status;
+	pid_t worker;
+
+	if (socketpair(PF_LOCAL, SOCK_SEQPACKET, 0, sk_pair) < 0) {
+		perror("socketpair");
+		return 1;
+	}
+	worker = fork();
+	if (worker < 0) {
+		perror("fork");
+		return 1;
+	}
+	if (worker == 0) {
+		close(sk_pair[0]);
+		ret = setup_fsfd_in_namespaces(sk_pair[1], pidfd);
+		close(sk_pair[1]);
+		exit(ret);
+	}
+	close(sk_pair[1]);
+	fsfd = recv_fd(sk_pair[0]);
+	assert(fsfd >= 0);
+	if (waitpid(worker, &status, 0) != worker) {
+		perror("waitpid");
+		return -1;
+	}
+	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+		fprintf(stderr, "worker exited nonzero\n");
+		return -1;
```

Comment:
why -1 in this case?

Reply:
Seems like if something went wrong, the fsfd we got may not be valid.
But it looks like we don't check for -1, eh?

Is your concern that we are returning sentinel values for no reason, or that the parent process shouldn't care?

Reply:
I have no idea what exit(-1) will do. Exit codes are supposed to be positive.

Reply:
Hmm, we were not exiting -1 from here, we were returning -1 as the fsfd.
I've changed it to exit 1, these are all fatal errors.